scripts/test/lint.py

```diff
@@ -0,0 +1,327 @@
+ is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and limitations under the License.
+"""
+import os
+import sys
+import logging
+import pathlib
+import argparse
+import datetime
+import subprocess
+from typing import Dict
+from xml.dom import minidom
+
+GIT_EXE = "git"
+
+# set log level for debugging here script-wide
+LOG_LEVEL = logging.INFO
+logging.basicConfig(level=LOG_LEVEL)
+logger = logging.getLogger("lint")
+
+
+class Lint:
+    WARN = "WARN"
+    FAIL = "FAIL"
+
+    name = "lint"
+    level = FAIL
+    recommendation = ""
+
+    def check(self, path):
+        """
+        :return: True if lint failed
+        """
+        return False
+
+
+def run_lints(lints, path):
+    for lint_ in lints:
+        if lint_.check(path):
+            yield lint_
+
+
+class FileNameNotAllLower(Lint):
+    name = "file name not all lower case"
+    recommendation = "Rename package path using all lower case letters"
+
+    def check(self, path):
+        return not path.name.islower()
+
+
+PATH_LINTS = (FileNameNotAllLower(),)
+
+
+def lint_path(path):
+    return run_lints(PATH_LINTS, path)
+
+
+def run_cmd(cmd):
+    logger.debug("cmd: %s", cmd)
+    p = subprocess.Popen(cmd.split(" "), stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    out, err = p.communicate()
+    out = out.decode("utf-8").strip()
+    err = err.decode("utf-8").strip()
+    logger.debug("cmd out: %s", out)
+    logger.debug("cmd err: %s", err)
+    return out
+
+
+class IncludesRequiredFieldsOnly(Lint):
+    name = "file lists non-required fields"
+    allowed_fields = [
+        "id",
+        "version",
+        "description",
+        "authors",
+        "dependencies",
+    ]
+    recommendation = f"Only include required fields: {', '.join(allowed_fields)}"
+
+    def check(self, path):
+        dom = minidom.parse(str(path))
+        metadata = dom.getElementsByTagName("metadata")[0]
+
+        non_required_fields = list()
+        for c in metadata.childNodes:
+            if c.nodeName == "#text":
+                continue
+            if c.nodeName not in self.allowed_fields:
+                non_required_fields.append(c.nodeName)
+        return len(non_required_fields) != 0
+
+
+class VersionFormatIncorrect(Lint):
+    name = "the version value is invalid"
+    recommendation = "see the Contributing Wiki for guidance on the version string"
+
+    def check(self, path):
+        dom = minidom.parse(str(path))
+        metadata = dom.getElementsByTagName("metadata")[0]
+
+        version = metadata.getElementsByTagName("version")[0]
+        version = version.firstChild.data.split(".")
+        if len(version) > 4:
+            print(f"{path} more than four version segments: {version}")
+            return True
+
+        if len(version) == 4:
+            try:
+                datetime.datetime.strptime(version[3], "%Y%m%d")
```

Comment:
we could also check that segments 1, 2 and 3 are a number

Reply:
100b7d6

Reply:
added, if we have to adjust later for `rc` or `beta` or whatever, it's an easy fix